consensus/src/dag/dag_fetcher.rs

```diff
@@ -103,17 +119,69 @@ impl DagFetcher {
                 .and_then(FetchResponse::try_from)
                 .and_then(|response| response.verify(&remote_request, &self.epoch_state.verifier))
             {
+                let ceritified_nodes = response.certified_nodes();
                 // TODO: support chunk response or fallback to state sync
-                let mut dag_writer = self.dag.write();
-                for rounds in response.certified_nodes() {
-                    for node in rounds {
+                {
+                    let mut dag_writer = self.dag.write();
+                    for node in ceritified_nodes {
                         if let Err(e) = dag_writer.add_node(node) {
                             error!("Failed to add node {}", e);
                         }
                     }
                 }
-                local_request.notify();
+
+                if self.dag.read().all_exists(local_request.node().parents()) {
+                    local_request.notify();
+                } else {
+                    // TODO: implement retry logic
+                }
             }
         }
     }
 }
+
+#[derive(Debug, ThisError)]
+pub enum FetchRequestHandleError {
+    #[error("parents are missing")]
+    ParentsMissing,
+}
+
+pub struct FetchRequestHandler {
+    dag: Arc<RwLock<Dag>>,
+}
+
+impl FetchRequestHandler {
+    pub fn new(dag: Arc<RwLock<Dag>>) -> Self {
+        Self { dag }
+    }
+}
+
+impl RpcHandler for FetchRequestHandler {
+    type Request = RemoteFetchRequest;
+    type Response = FetchResponse;
+
+    fn process(&mut self, message: Self::Request) -> anyhow::Result<Self::Response> {
+        let dag_reader = self.dag.read();
+
+        ensure!(
+            message
+                .parents()
+                .iter()
+                .all(|metadata| dag_reader.get_node(metadata).is_some()),
```

Comment:
looks like redundant to all_exists? probably can take impl Iterator<&NodeMetadata> and having a function to return this from certified_node.parents too